// conventional-component/dist/actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.destroy = exports.receiveNextProps = exports.init = exports.DESTROY = exports.RECEIVE_NEXT_PROPS = exports.INIT = undefined;

var _withActionIdentity = require('./withActionIdentity');

var _withActionIdentity2 = _interopRequireDefault(_withActionIdentity);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var INIT = 'conventional-component/INIT';
var RECEIVE_NEXT_PROPS = 'conventional-component/RECEIVE_NEXT_PROPS';
var DESTROY = 'conventional-component/DESTROY';

var init = (0, _withActionIdentity2.default)(function () {
  var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return {
    type: INIT,
    payload: props
  };
});

var receiveNextProps = (0, _withActionIdentity2.default)(function () {
  var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return {
    type: RECEIVE_NEXT_PROPS,
    payload: props
  };
});

var destroy = (0, _withActionIdentity2.default)(function () {
  var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return {
    type: DESTROY
  };
});

exports.INIT = INIT;
exports.RECEIVE_NEXT_PROPS = RECEIVE_NEXT_PROPS;
exports.DESTROY = DESTROY;
exports.init = init;
exports.receiveNextProps = receiveNextProps;
exports.destroy = destroy;
